File: ta4j-core/src/test/java/org/ta4j/core/indicators/numeric/statistics/CorrelationCoefficientIndicatorTest.java
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2024 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.indicators.numeric.statistics;

import static org.ta4j.core.TestUtils.assertNext;
import static org.ta4j.core.TestUtils.fastForward;

import org.junit.Before;
import org.junit.Test;
import org.ta4j.core.MockStrategy;
import org.ta4j.core.backtest.BacktestBarSeries;
import org.ta4j.core.indicators.AbstractIndicatorTest;
import org.ta4j.core.indicators.numeric.NumericIndicator;
import org.ta4j.core.mocks.MockBarSeriesBuilder;
import org.ta4j.core.num.Num;
import org.ta4j.core.num.NumFactory;

public class CorrelationCoefficientIndicatorTest extends AbstractIndicatorTest<Num> {

  private NumericIndicator close;
  private NumericIndicator volume;
  private BacktestBarSeries data;


  public CorrelationCoefficientIndicatorTest(final NumFactory numFactory) {
    super(numFactory);
  }


  @Before
  public void setUp() {
    this.data = new MockBarSeriesBuilder().withNumFactory(this.numFactory).build();
    this.data.barBuilder().closePrice(6).volume(100).add();
    this.data.barBuilder().closePrice(7).volume(105).add();
    this.data.barBuilder().closePrice(9).volume(130).add();
    this.data.barBuilder().closePrice(12).volume(160).add();
    this.data.barBuilder().closePrice(11).volume(150).add();
    this.data.barBuilder().closePrice(10).volume(130).add();
    this.data.barBuilder().closePrice(11).volume(95).add();
    this.data.barBuilder().closePrice(13).volume(120).add();
    this.data.barBuilder().closePrice(15).volume(180).add();
    this.data.barBuilder().closePrice(12).volume(160).add();
    this.data.barBuilder().closePrice(8).volume(150).add();
    this.data.barBuilder().closePrice(4).volume(200).add();
    this.data.barBuilder().closePrice(3).volume(150).add();
    this.data.barBuilder().closePrice(4).volume(85).add();
    this.data.barBuilder().closePrice(3).volume(70).add();
    this.data.barBuilder().closePrice(5).volume(90).add();
    this.data.barBuilder().closePrice(8).volume(100).add();
    this.data.barBuilder().closePrice(9).volume(95).add();
    this.data.barBuilder().closePrice(11).volume(110).add();
    this.data.barBuilder().closePrice(10).volume(95).add();

    this.close = NumericIndicator.closePrice(this.data);
    this.volume = NumericIndicator.volume(this.data);
  }


  @Test
  public void usingBarCount5UsingClosePriceAndVolume() {
    final var coef = new CorrelationCoefficientIndicator(this.close, this.volume, 5);
    this.data.replaceStrategy(new MockStrategy(coef));

    fastForward(this.data, 5);
    assertNext(this.data, 0.7977, coef);
    assertNext(this.data, 0.7842, coef);
    assertNext(this.data, 0.2536, coef);
    assertNext(this.data, 0.0548, coef);
    assertNext(this.data, 0.4841, coef);
    assertNext(this.data, 0.5219, coef);
    assertNext(this.data, 0.2436, coef);
    assertNext(this.data, -0.3662, coef);
    assertNext(this.data, 0.02518, coef);
    assertNext(this.data, 0.1169, coef);
    assertNext(this.data, 0.2139, coef);
    assertNext(this.data, -0.0481, coef);
    assertNext(this.data, -0.0853, coef);
    assertNext(this.data, 0.6913, coef);
    assertNext(this.data, 0.7473, coef);
    assertNext(this.data, 0.5843, coef);
  }
}

File: ta4j-core/src/test/java/org/ta4j/core/indicators/numeric/statistics/CorrelationCoefficientIndicatorTest.kt
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2024 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.indicators.numeric.statistics

import java.time.Duration
import java.time.Instant
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.MethodSource
import org.ta4j.core.MarketEventTestContext
import org.ta4j.core.api.Indicators.closePrice
import org.ta4j.core.events.CandleReceived
import org.ta4j.core.indicators.TimeFrame
import org.ta4j.core.indicators.numeric.candles.VolumeIndicator
import org.ta4j.core.num.NumFactory

class CorrelationCoefficientIndicatorTest {

    private fun createCandleEvents(data: List<Pair<Double, Double>>): List<CandleReceived> {
        return data.mapIndexed { index, (close, volume) ->
            CandleReceived(
                timeFrame = TimeFrame.DAY,
                beginTime = Instant.EPOCH.plus(Duration.ofDays(index.toLong())),
                endTime = Instant.EPOCH.plus(Duration.ofDays(index + 1L)),
                closePrice = close,
                volume = volume
            )
        }
    }

    @ParameterizedTest
    @MethodSource("org.ta4j.core.NumFactoryTestSource#numFactories")
    fun usingBarCount5UsingClosePriceAndVolume(numFactory: NumFactory) {
        val testData = listOf(
            6.0 to 100.0, 7.0 to 105.0, 9.0 to 130.0, 12.0 to 160.0, 11.0 to 150.0,
            10.0 to 130.0, 11.0 to 95.0, 13.0 to 120.0, 15.0 to 180.0, 12.0 to 160.0,
            8.0 to 150.0, 4.0 to 200.0, 3.0 to 150.0, 4.0 to 85.0, 3.0 to 70.0,
            5.0 to 90.0, 8.0 to 100.0, 9.0 to 95.0, 11.0 to 110.0, 10.0 to 95.0
        )

        val context = MarketEventTestContext()
            .withNumFactory(numFactory)
            .withMarketEvents(createCandleEvents(testData))

        val coef = CorrelationCoefficientIndicator(closePrice(), VolumeIndicator(numFactory), 5)

        context.withIndicator(coef)
            .fastForwardUntilStable()
            .assertCurrent(0.7977)
            .assertNext(0.7842)
            .assertNext(0.2536)
            .assertNext(0.0548)
            .assertNext(0.4841)
            .assertNext(0.5219)
            .assertNext(0.2436)
            .assertNext(-0.3662)
            .assertNext(0.02518)
            .assertNext(0.1169)
            .assertNext(0.2139)
            .assertNext(-0.0481)
            .assertNext(-0.0853)
            .assertNext(0.6913)
            .assertNext(0.7473)
            .assertNext(0.5843)
    }
}
